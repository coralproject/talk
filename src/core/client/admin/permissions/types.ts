import { GQLUSER_ROLE_RL } from "coral-framework/schema";

export type PermissionMap<AbilityType extends string, ContextType> = {
  [abilityType in AbilityType]: {
    [role in GQLUSER_ROLE_RL]?: (context?: ContextType) => boolean;
  };
};
