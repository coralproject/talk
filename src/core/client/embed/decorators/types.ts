import pym from "pym.js";

export type CleanupCallback = () => void;
export type Decorator = (pym: pym.Parent) => CleanupCallback | void;
