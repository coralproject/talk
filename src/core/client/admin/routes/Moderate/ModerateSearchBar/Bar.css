@import "variables.css";

.root {
  height: calc(6 * var(--mini-unit));
  background-color: $story-search-background;
  margin-top: -1px;
  border-bottom: 0;
}

.bumpZIndex {
  z-index: 100;
}

.popover {
  width: calc(84 * var(--mini-unit));
  border: 0;
}

.listBox {
  margin: 0;
  padding: 0;
  list-style: none;
}

.groupIcon {
  margin-right: var(--v2-spacing-2);
}
