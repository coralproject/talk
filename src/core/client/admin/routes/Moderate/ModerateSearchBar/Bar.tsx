import { Localized } from "@fluent/react/compat";
import key from "keymaster";
import React, {
  FunctionComponent,
  useCallback,
  useEffect,
  useRef,
} from "react";
import { Form } from "react-final-form";

import { HOTKEYS } from "coral-admin/constants";
import { Backdrop, Icon, Popover, SubBar } from "coral-ui/components/v2";
import { combineEventHandlers } from "coral-ui/helpers";
import {
  useBlurOnEsc,
  useComboBox,
  useFocus,
  usePreventFocusLoss,
} from "coral-ui/hooks";
import { ListBoxOption } from "coral-ui/hooks/useComboBox";

import Field from "./Field";
import Group from "./Group";

import styles from "./Bar.css";

/** Group of listbox options. */
type Group = "CONTEXT" | "SEARCH";

interface Props {
  /** title of the current story */
  title: string;
  /** options to show in the combobox listbox */
  options: Array<ListBoxOption & { group: Group }>;
  /** onSearch will be called whenenver the user submits the search */
  onSearch?: (value: string) => void;

  siteSelector: React.ReactNode;
}

/**
 * Bar is the container of the whole search bar.
 */
const Bar: FunctionComponent<Props> = ({
  title,
  options,
  onSearch,
  siteSelector,
}) => {
  const [focused, focusHandlers] = useFocus();
  const preventFocusLossHandlers = usePreventFocusLoss(focused);
  const submitHandler = useCallback(
    ({ search }: { search: string }) => onSearch && onSearch(search),
    [onSearch]
  );
  const searchInput = useRef<HTMLInputElement>(null);
  const blurOnEscProps = useBlurOnEsc(focused);
  const [
    mappedOptions,
    activeDescendant,
    keyboardNavigationHandlers,
  ] = useComboBox("moderate-searchBar-listBoxOption", options);

  useEffect(() => {
    key(HOTKEYS.SEARCH, () => {
      if (searchInput && searchInput.current) {
        searchInput.current.focus();
      }
    });
  }, [searchInput.current]);
  const contextOptions = mappedOptions
    .filter(o => o.group === "CONTEXT")
    .map(o => o.element);
  const searchOptions = mappedOptions
    .filter(o => o.group === "SEARCH")
    .map(o => o.element);

  return (
    <Localized id="moderate-searchBar-comboBox" attrs={{ "aria-label": true }}>
      <SubBar
        className={styles.root}
        data-testid="moderate-searchBar-container"
        role="combobox"
        aria-owns="moderate-searchBar-listBox"
        aria-label="Search or jump to story"
        aria-haspopup="listbox"
        aria-expanded={focused}
      >
        <Backdrop className={styles.bumpZIndex} active={focused} />
        {siteSelector}
        <Form onSubmit={submitHandler}>
          {({ handleSubmit }) => (
            <Localized
              id="moderate-searchBar-searchForm"
              attrs={{ "aria-label": true }}
            >
              <form
                role="search"
                aria-label="Stories"
                className={styles.bumpZIndex}
                onSubmit={handleSubmit}
                {...preventFocusLossHandlers}
              >
                <Popover
                  id={"moderate-searchBar-popover"}
                  placement="bottom"
                  classes={{ popover: styles.popover }}
                  visible={focused}
                  eventsEnabled={false}
                  modifiers={{
                    preventOverflow: { enabled: false },
                    flip: { enabled: false },
                    hide: { enabled: false },
                    arrow: { enabled: false },
                  }}
                  body={() => (
                    <ul
                      id="moderate-searchBar-listBox"
                      role="listbox"
                      className={styles.listBox}
                    >
                      {contextOptions.length > 0 && (
                        <Localized
                          id="moderate-searchBar-currentlyModerating"
                          attrs={{ title: true }}
                        >
                          <Group
                            title="Currently moderating"
                            id="moderate-searchBar-context"
                          >
                            {contextOptions}
                          </Group>
                        </Localized>
                      )}
                      {searchOptions.length === 1 && (
                        <Group
                          title={
                            <>
                              <Icon size="md" className={styles.groupIcon}>
                                search
                              </Icon>{" "}
                              <Localized id="moderate-searchBar-searchResults">
                                <span>Search results</span>
                              </Localized>
                            </>
                          }
                          id="moderate-searchBar-search"
                          light
                        >
                          {searchOptions}
                        </Group>
                      )}
                      {searchOptions.length > 1 && (
                        <Group
                          title={
                            <>
                              <Icon size="md" className={styles.groupIcon}>
                                search
                              </Icon>{" "}
                              <Localized id="moderate-searchBar-searchResultsMostRecentFirst">
                                <span>Search results (Most recent first)</span>
                              </Localized>
                            </>
                          }
                          id="moderate-searchBar-search"
                          light
                        >
                          {searchOptions}
                        </Group>
                      )}
                    </ul>
                  )}
                >
                  {({ ref }) => (
                    <div ref={ref}>
                      <Field
                        title={title}
                        ref={searchInput}
                        {...combineEventHandlers(
                          focusHandlers,
                          blurOnEscProps,
                          keyboardNavigationHandlers
                        )}
                        focused={focused}
                        aria-controls="moderate-searchBar-listBox"
                        aria-autocomplete="list"
                        aria-activedescendant={activeDescendant}
                      />
                    </div>
                  )}
                </Popover>
              </form>
            </Localized>
          )}
        </Form>
      </SubBar>
    </Localized>
  );
};

export default Bar;
