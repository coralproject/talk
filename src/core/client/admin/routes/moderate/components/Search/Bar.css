.root {
  height: calc(6 * var(--mini-unit));
  background-color: #013f68;
  margin-top: -1px;
}

.bumpZIndex {
  z-index: 100;
}

.popover {
  width: calc(94 * var(--mini-unit));
  border: 0;
}

.listBox {
  margin: 0;
  padding: 0;
  list-style: none;
}
