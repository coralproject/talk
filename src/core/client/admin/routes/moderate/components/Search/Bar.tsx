import { Localized } from "fluent-react/compat";
import React, { FunctionComponent, useCallback } from "react";
import { Form } from "react-final-form";

import { Backdrop, Icon, Popover, SubBar } from "talk-ui/components";
import { combineEventHandlers } from "talk-ui/helpers";
import {
  useBlurOnEsc,
  useComboBox,
  useFocus,
  usePreventFocusLoss,
} from "talk-ui/hooks";
import { ListBoxOption } from "talk-ui/hooks/useComboBox";

import Field from "./Field";
import Group from "./Group";

import styles from "./Bar.css";

/** Group of listbox options. */
type Group = "CONTEXT" | "SEARCH";

interface Props {
  /** title of the current story */
  title: string;
  /** options to show in the combobox listbox */
  options: Array<ListBoxOption & { group: Group }>;
  /** onSearch will be called whenenver the user submits the search */
  onSearch?: (value: string) => void;
}

/**
 * Bar is the container of the whole search bar.
 */
const Bar: FunctionComponent<Props> = ({ title, options, onSearch }) => {
  const [focused, focusHandlers] = useFocus();
  const preventFocusLossHandlers = usePreventFocusLoss(focused);
  const submitHandler = useCallback(
    ({ search }: { search: string }) => onSearch && onSearch(search),
    [onSearch]
  );
  const blurOnEscProps = useBlurOnEsc(focused);
  const [
    mappedOptions,
    activeDescendant,
    keyboardNavigationHandlers,
  ] = useComboBox("moderate-searchBar-listBoxOption", options);

  const contextOptions = mappedOptions
    .filter(o => o.group === "CONTEXT")
    .map(o => o.element);
  const searchOptions = mappedOptions
    .filter(o => o.group === "SEARCH")
    .map(o => o.element);

  return (
    <Localized id="moderate-searchBar-comboBox" attrs={{ "aria-label": true }}>
      <SubBar
        className={styles.root}
        data-testid="moderate-searchBar-container"
        role="combobox"
        aria-owns="moderate-searchBar-listBox"
        aria-label="Search or jump to story"
        aria-haspopup="listbox"
        aria-expanded={focused}
      >
        <Backdrop className={styles.bumpZIndex} active={focused} />
        <Form onSubmit={submitHandler}>
          {({ handleSubmit }) => (
            <Localized
              id="moderate-searchBar-searchForm"
              attrs={{ "aria-label": true }}
            >
              <form
                role="search"
                aria-label="Stories"
                className={styles.bumpZIndex}
                onSubmit={handleSubmit}
                {...preventFocusLossHandlers}
              >
                <Popover
                  id={"moderate-searchBar-popover"}
                  placement="bottom"
                  classes={{ popover: styles.popover }}
                  visible={focused}
                  eventsEnabled={false}
                  modifiers={{
                    preventOverflow: { enabled: false },
                    flip: { enabled: false },
                    hide: { enabled: false },
                  }}
                  body={() => (
                    <ul
                      id="moderate-searchBar-listBox"
                      role="listbox"
                      className={styles.listBox}
                    >
                      {contextOptions.length > 0 && (
                        <Localized
                          id="moderate-searchBar-currentlyModerating"
                          attrs={{ title: true }}
                        >
                          <Group
                            title="Currently moderating"
                            id="moderate-searchBar-context"
                          >
                            {contextOptions}
                          </Group>
                        </Localized>
                      )}
                      {searchOptions.length > 0 && (
                        <Group
                          title={
                            <>
                              <Icon>search</Icon>{" "}
                              <Localized id="moderate-searchBar-searchResultsMostRecentFirst">
                                <span>Search results (Most recent first)</span>
                              </Localized>
                            </>
                          }
                          id="moderate-searchBar-search"
                          light
                        >
                          {searchOptions}
                        </Group>
                      )}
                    </ul>
                  )}
                >
                  {({ ref }) => (
                    <div ref={ref}>
                      <Field
                        title={title}
                        {...combineEventHandlers(
                          focusHandlers,
                          blurOnEscProps,
                          keyboardNavigationHandlers
                        )}
                        focused={focused}
                        aria-controls="moderate-searchBar-listBox"
                        aria-autocomplete="list"
                        aria-activedescendant={activeDescendant}
                      />
                    </div>
                  )}
                </Popover>
              </form>
            </Localized>
          )}
        </Form>
      </SubBar>
    </Localized>
  );
};

export default Bar;
