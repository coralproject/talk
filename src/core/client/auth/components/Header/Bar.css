.root {
  min-height: calc(var(--mini-unit) * 8);
  padding: var(--mini-unit);
  background-color: var(--palette-text-500);
  box-sizing: border-box;
}
