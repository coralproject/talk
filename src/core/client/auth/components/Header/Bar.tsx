import React, { FunctionComponent } from "react";

import { Flex } from "coral-ui/components";

import styles from "./Bar.css";

export interface BarProps {
  children: React.ReactNode;
}

const Bar: FunctionComponent<BarProps> = props => (
  <Flex className={styles.root} alignItems="center" justifyContent="center">
    <header>{props.children}</header>
  </Flex>
);

export default Bar;
