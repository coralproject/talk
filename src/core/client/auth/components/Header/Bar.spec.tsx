import React from "react";
import { createRenderer } from "react-test-renderer/shallow";

import { PropTypesOf } from "talk-framework/types";

import Bar from "./Bar";

it("renders correctly", () => {
  const props: PropTypesOf<typeof Bar> = {
    children: "Hello World",
  };
  const renderer = createRenderer();
  renderer.render(<Bar {...props} />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
