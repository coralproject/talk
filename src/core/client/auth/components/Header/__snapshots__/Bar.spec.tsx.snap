// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<ForwardRef(forwardRef)
  alignItems="center"
  className="Bar-root"
  justifyContent="center"
>
  <header>
    Hello World
  </header>
</ForwardRef(forwardRef)>
`;
